### PR TITLE
=str Add IterableSource.

### PR DESCRIPTION
**Pull request by He-Pin**
was https://github.com/akka/akka/pull/31372

This introduce a dedicated IterableSource, which can be later used for performance optimization. And also obey the supervisionStrategy now.

### CODE REVIEW

File: stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/SourceSpec.scala
Comment (pjfanning) [inline stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/SourceSpec.scala:351]: `shouldBe a[TE]` would be better

**Review by pjfanning — APPROVED** [review]
lgtm - but I'd prefer if some other reviewers also had a look before we merge

File: stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/SourceSpec.scala
Comment (nvollmar) [inline stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/SourceSpec.scala:314]: I'd prefer to have separate tests with descriptive titles if they test different edge cases, or at least a comment per run hinting what each is testing.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala
Comment (nvollmar) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:65]: This check only makes sense if we pushed an element -> move it inside the `if (isAvailable(out)) {` block

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala
Comment (nvollmar) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:45]: couldn't we just initialize the variable with `elements.iterator` and avoid the if check/assignment here?

Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:45]: There are some cases where the `iterator` method call invoving some underling resources creation, that's why I defer the creation in the first place.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala
Comment (nvollmar) [inline stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:60]: That would even be simpler:

```
      private def tryPushNextOrComplete(): Unit = {
         if (currentIterator.hasNext && isAvailable(out)) {
           push(out, currentIterator.next())
         }

         if (!currentIterator.hasNext) {
           completeStage()
         }
      }
```

Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:60]: But this will do another `currentIterator.hasNext` checking.

File: stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/fusing/IterableSource.scala:45]: IIRC the JOOQ does this, and our internal Lindorm Nosql db driver does the same too.

**Pull request merged**
